doc/man/man1/context.1
\"    This documentation is copyright 1994 David Huron.
.TH context 1 "1994 Dec. 4"
.AT 3
.sp 2
.SH "NAME"
.in +2
.in +13
.ti -13
\fBcontext\fR  --  congeal Humdrum data records to form a contextual frame
.in -13
.in -2
.sp 1
.sp 1
.SH "SYNOPSIS"
.in +2
\fBcontext\fR [\fB-b \fIregexp\fR] [\fB-d \fIstring\fR] [\fB-e \fIregexp\fR] [\fB-i \fIregexp\fR] [\fB-n \fIn\fR] [\fB-o \fIregexp\fR] [\fB-p \fIn\fR] [\fIinputfile\fR ...]
.in -2
.sp 1
.sp 1
.SH "DESCRIPTION"
.in +2
The
.B "context"
command amalgamates one or more successive input data records into
single records according to user-defined criteria.
Only single-spine Humdrum inputs are permitted.
The
.B "context"
command provides a useful means for amalgamating on a single line
those data tokens that are somehow deemed to be contextually related.
For example,
.B "context"
might be used to link together all pitches in a measure,
or pair the first and last notes of each phrase.
The command is useful in such tasks as partitioning possible pitch-class sets
or grouping arpeggio tones into chords.
.sp 1
.sp 1
In its simplest mode of operation
.B "context"
will join a specified number of successive data records together
to form a single output record.
By way of example, consider a file (named \f(CRinput\fR) consisting of a single
spine whose data records contain the numbers 1 through 6 on separate lines:
.in +2
.sp 1
.TS
l l l l l.
**numbers
1
2
3
4
5
6
*-
.TE
.sp 1
.in -2
The command
.sp 1
.sp 1
.in +2
context -n 3 input
.in -2
.sp 1
.sp 1
will produce the following output:
.in +2
.sp 1
.TS
l.
**numbers
1 2 3
2 3 4
3 4 5
4 5 6
\.
\.
*-
.TE
.sp 1
.in -2
Notice that the output file has been padded with null tokens so that
the number of output records matches the number of input records.
By invoking the
.B "-p"
option, the padded null tokens can be placed either at the beginning or the
end of the file, or split between beginning and end.
For example, the command
.sp 1
.sp 1
.in +2
context -n 3 -p 3 input
.in -2
.sp 1
.sp 1
will place the trailing null tokens in the above example at the beginning
of the output.
The
.B "-p"
option defaults to the value 0.
.sp 1
.sp 1
If null tokens are present in the input, they remain in place,
yet do not affect the congealed data records.
For example, if a null token was present between the
numbers 1 and 2 in the above input, the command
.sp 1
.sp 1
.in +2
context -n 4 -p 1 input
.in -2
.sp 1
.sp 1
would produce an output beginning with a single padded null token:
.in +2
.sp 1
.TS
l.
**numbers
\.
1 2 3 4
\.
2 3 4 5
3 4 5 6
\.
\.
*-
.TE
.sp 1
.in -2
Rather than specifying a fixed number of congealed data records,
input records can alternatively be amalgamated according to the
signifiers present in the input data itself.
The
.B "-e"
option allows the user to specify an \(odend\(cd signifier.
When this signifier is encountered in the input,
the input record is appended to the current congealed record
-- which is then output -- and a new congealed output record begins.
End signifiers are defined as string patterns using the
.I "regular expression"
syntax (see \fBregexp (5)\fR).
For example, given an input of six successive numbers, the command
.sp 1
.sp 1
.in +2
context -e [0246] input
.in -2
.sp 1
.sp 1
would produce the following output:
.in +2
.sp 1
.TS
l.
**numbers
1 2
\.
3 4
\.
5 6
\.
*-
.TE
.sp 1
.in -2
.in -2
.SH "OPTIONS"
.in +2
The
.B "context"
command supports the following options:
.sp 1
.TS
l l.
\fB-b \fIregexp\fR	begin a new output record starting with token matching
	\fIregexp\fR
\fB-d \fIstring\fR	use \fIstring\fR as output delimiter for input records
	rather than the space character
\fB-e \fIregexp\fR	begin a new output record starting after token matching
	\fIregexp\fR
\fB-h\fR	displays a help screen summarizing the command syntax
\fB-i \fIregexp\fR	ignore any records matching \fIregexp\fR when counting
\fB-n \fIn\fR	amalgamate \fIn\fR input data records for each output record
\fB-o \fIregexp\fR	omit any records matching \fIregexp\fR from amalgamated
	output; do not count
\fB-p \fIn\fR	pad \fIn\fR (normally trailing) null tokens at the beginning of
	the output spine
.TE
.sp 1
Options are specified in the command line.
Note that the \fB-b\fR and \fB-e\fR options are mutually exclusive
with the \fB-n\fR, \fB-p\fR, and \fB-i\fR options.
.sp 1
.sp 1
In the default operation,
.B "context"
separates amalgamated tokens by inserting a space character.
(Thus the input tokens are treated as subtokens in a Humdrum multiple-stop.)
The
.B "-d"
option allows the user to define an alternative string as the
subtoken delimiter.
.sp 1
.sp 1
The
.B "-n"
option allows the user to specify the maximum number of data records
assembled into a single output record.
.sp 1
.sp 1
The
.B "-b"
option allows the user to specify a \(odbegin\(cd marker.
When this marker (\fIregexp\fR) is matched in the input,
any current congealed record is output, and a new congealed
record begins.
Begin signifiers are defined as string patterns using the
.I "regular expression"
syntax.
.sp 1
.sp 1
With the
.B "-e"
option, if
.B "context"
encounters a data record matching
.I "regexp"
then it appends the current input record to the current assembled output
record and begins assembling a new record with the following input record.
.sp 1
.sp 1
The
.B "-i"
option is used only with
.B "-n;"
it causes any data records matching
.I "regexp"
to be ignored in the counting of amalgamated tokens.
Such \(oduncounted\(cd records are nevertheless output.
.sp 1
.sp 1
The
.B "-o"
option causes data records matching
.I "regexp"
to be omitted from the output.
.sp 1
.sp 1
The
.B "-p"
option may be used in conjunction with \fB-n\fR.
Normally, the output from
.B "context -n"
is padded with trailing null tokens -- one fewer in number than
the value specified with \fB-n\fR.
The \fB-p \fIn\fR option causes \fIn\fR null tokens to be padded at the
.I "beginning"
of the output spine, rather than trailing at the end.
.sp 1
.sp 1
Note that tandem interpretations and comments are processed like null tokens;
they are merely echoed in the output in their appropriate position.
Note also that
.B "context"
automatically breaks a congealed output record whenever
it encounters a spine-path terminator or exclusive interpretation
in the input.
.in -2
.sp 1
.sp 1
.SH "EXAMPLES"
.in +2
The following excerpt from Edgar Var\o'e\(ga'se's
.I "Density 21.5 (1936)"
illustrates the use of
.B "context."
Consider the initial input:
.in +2
.sp 1
.TS
l.
!! Edgar Var\o'e\(ga'se, Density 21.5 (1936)
!! excerpt: mm.41-45
**kern
*MM72
\(eq41
(16f#
16e#
[8gn
2.g_
\(eq42
4g])
4r
4r
8r
(16f#
16e#
\(eq43
6gn)
(6e#
6f#
8g)
(8f#
12e#
12g
12dn
\(eq44
2a-)
(4an
8een
[8bb-
\(eq45
4bb-]
2.ccc#)
\(eq45
8r
*-
.TE
.sp 1
.in -2
A simple transformation would be to amalgamate successive data
records in overlapping groups of 3.
The following command:
.sp 1
.sp 1
.in +2
context -n 3 density
.in -2
.sp 1
.sp 1
would produce the following output:
.in +2
.sp 1
.TS
l.
!! Edgar Var\o'e\(ga'se, Density 21.5 (1936)
!! excerpt: mm.41-45
**kern
*MM72
\(eq41 (16f# 16e#
(16f# 16e# [8gn
16e# [8gn 2.g_
[8gn 2.g_ \(eq42
2.g_ \(eq42 4g])
\(eq42 4g]) 4r
4g]) 4r 4r
4r 4r 8r
4r 8r (16f#
8r (16f# 16e#
(16f# 16e# \(eq43
16e# \(eq43 6gn)
\(eq43 6gn) (6e#
6gn) (6e# 6f#
(6e# 6f# 8g)
6f# 8g) (8f#
8g) (8f# 12e#
(8f# 12e# 12g
12e# 12g 12dn
12g 12dn \(eq44
12dn \(eq44 2a-)
\(eq44 2a-) (4an
2a-) (4an 8een
(4an 8een [8bb-
8een [8bb- \(eq45
[8bb- \(eq45 4bb-]
\(eq45 4bb-] 2.ccc#)
4bb-] 2.ccc#) \(eq45
2.ccc#) \(eq45 8r
\.
\.
*-
.TE
.sp 1
.in -2
Notice once again that the input and output have the same number of records.
Preserving the structure in this way allows the user to
coordinate the contextual output with the original input using the
.B "assemble"
command.
.sp 1
.sp 1
A more useful transformation might amalgamate successive data
records in overlapping groups of 3
.I "notes;"
that is rests and barlines should be ignored.
The following command causes input records containing either an equals-sign
or the letter `r' to be ignored when counting the number of amalgamated
data records:
.sp 1
.sp 1
.in +2
context -n 3 -i [=r] density
.in -2
.sp 1
.sp 1
The input and corresponding output are given in the left
and right spines below:
.in +2
.sp 1
.TS
l s
l s
l l.
!! Edgar Var\o'e\(ga'se, Density 21.5 (1936)
!! excerpt: mm.41-45
**kern	**kern
*MM72	*MM72
\(eq41	\(eq41 (16f# 16e# [8gn
(16f#	(16f# 16e# [8gn
16e#	16e# [8gn 2.g_
[8gn	[8gn 2.g_ \(eq42 4g])
2.g_	2.g_ \(eq42 4g]) 4r 4r 8r (16f#
\(eq42	\(eq42 4g]) 4r 4r 8r (16f# 16e#
4g])	4g]) 4r 4r 8r (16f# 16e#
4r	4r 4r 8r (16f# 16e# \(eq43 6gn)
4r	4r 8r (16f# 16e# \(eq43 6gn)
8r	8r (16f# 16e# \(eq43 6gn)
(16f#	(16f# 16e# \(eq43 6gn)
16e#	16e# \(eq43 6gn) (6e#
\(eq43	\(eq43 6gn) (6e# 6f#
6gn)	6gn) (6e# 6f#
(6e#	(6e# 6f# 8g)
6f#	6f# 8g) (8f#
8g)	8g) (8f# 12e#
(8f#	(8f# 12e# 12g
12e#	12e# 12g 12dn
12g	12g 12dn \(eq44 2a-)
12dn	12dn \(eq44 2a-) (4an
\(eq44	\(eq44 2a-) (4an 8een
2a-)	2a-) (4an 8een
(4an	(4an 8een [8bb-
8een	8een [8bb- \(eq45 4bb-]
[8bb-	[8bb- \(eq45 4bb-] 2.ccc#)
\(eq45	.
4bb-]	.
2.ccc#)	.
\(eq45	.
8r	.
*-	*-
.TE
.sp 1
.in -2
Notice that as the end of the file is approached,
.B "context"
will continue amalgamating data records until it
is no longer able to satisfy the amalgamating criteria.
If unable to complete an output record,
.B "context"
will output a null token.
.sp 1
.sp 1
If the above command had used the
.B "-o"
rather than the
.B "-i"
option, all of the rests and barlines would have been omitted
from the output.
Otherwise, the output would be the same as given above.
.sp 1
.sp 1
A more musically useful partitioning of Var\o'e\(ga'se's
work might be based on slur markings.
The following command uses open- and closed-slur markings
to demarcate the contextual outputs;
(note the need to escape the parentheses since they are regular expression
metacharacters).
.sp 1
.sp 1
.in +2
context -b '\\(' -e '\\)' -o '[=r]' density
.in -2
.sp 1
.sp 1
Notice that the ensuing output (shown below) pads the output with
null tokens in order to maintain the same
number of data records as the original input.
(An output such as the following might be used as input to
a command sequence such as \fBpc -x | pcset\fR).
.in +2
.sp 1
.TS
l.
!! Edgar Var\o'e\(ga'se, Density 21.5 (1936)
!! excerpt: mm.41-45
**kern
*MM72
\.
(16f# 16e# [8gn 2.g_ 4g])
\.
\.
\.
\.
\.
\.
\.
\.
(16f# 16e# 6gn)
\.
\.
\.
(6e# 6f# 8g)
\.
\.
(8f# 12e# 12g 12dn 2a-)
\.
\.
\.
\.
\.
(4an 8een [8bb- 4bb-] 2.ccc#)
\.
\.
\.
\.
\.
\.
\.
*-
.TE
.sp 1
.in -2
If there were any notes not embraced within a slur,
the above command would have caused them to be output on
their own line.
.in -2
.sp 1
.sp 1
.SH "PORTABILITY"
.in +2
\s-1DOS\s+1 2.0 and up, with the \s-1MKS\s+1 Toolkit.
\s-1OS/2\s+1 with the \s-1MKS\s+1 Toolkit.
\s-1UNIX\s+1 systems supporting the
.I "Korn"
shell or
.I "Bourne"
shell command interpreters, and revised
.I "awk"
(1985).
.in -2
.sp 1
.sp 1
.SH "SEE ALSO"
.in +2
\fBcontext\fR (1), \fBgrep\fR (\s-1UNIX\s+1),
\fBnf\fR (1), \fBpatt\fR (1),
\fBpattern\fR (1), \fBpcset\fR (1)
.in -2
.sp 1
.sp 1
.SH "WARNINGS"
.in +2
The \fB-b\fR and \fB-e\fR options are mutually exclusive
with the \fB-n\fR, \fB-p\fR, and \fB-i\fR options.
.in -2


toolkit/awk/context.awk
#####################################################################
##                          CONTEXT.AWK
# Programmed by:  Tim Racinsky     Date: May 25/94
# Copyright (c) 1994 David Huron
#
# Modifications:
#   Date:      Programmer:    Description:
# 1997 Mar 18  David Huron    Fixed bug with -o option
#
#	This program congeals data records to form a contextual frame
#
# Main Functions used:
#	parse_command()		process_terminate()		process_echo()
#	process_data()			flush_buffer()			process_noption()
#	process_boption()		process_eoption()		print_buffer()
#	reset_new()			reset_space()
#
#	VARIABLES:
#
#  -stderr:  Destination of error messages
#  -echo:  Regular expression to match comments, tand. interps., and null tokens
#  -buffer_head:  holds position of last entry in array buffer
#  -buffer_tail:  holds postion of oldest non-printed entry in buffer
#  -data_count:  holds number of 'counted' data records not printed in buffer
#  -storing:  flag to indicate if currently storing records in buffer
#  -printed_pre_nulls:  flag to indicate if pre-nulls have been printed for -p
#  -first_file:  flag to indicate if this is the first file processed
#  -options:  holds options specified by user
#  -begin_reg:  holds regular expression for -b option
#  -end_reg:  holds regular expression for -e option
#  -delimiter:  holds string value for -d option
#  -del_length:  holds length of string for -d option
#  -ignore_reg:  holds regular expression for -i option
#  -omit_reg:  holds regular expression for -o option
#  -number:   holds value for -n option
#  -placement:   holds value for -p option
#
#	ARRAYS:
#
#  -buffer:  Holds data records temporarily so they can be processed
#            several times
#
BEGIN {
	USAGE="\nUSAGE: context -h                   (Help Screen)\n       contex"\
		 "t [-b regexp] [-d string] [-e regexp] [-i regexp] [-n n]\n       "\
		 "        [-o regexp] [-p n] [file ...]"
	FS = OFS = "\t"
	TRUE = 1; FALSE = 0
	#
	# Set standard error to the appropriate 'file'
	#
	#if (ARGV[0] == "nawk") stderr = "/dev/tty"
	#else stderr = "con"
	"echo $CON" | getline stderr
	close("echo $CON")
	#
	# Set other global variables
	#
	echo = "^(([!*])|(\\.$))"
	buffer_head = 0
	buffer_tail = 1
	data_count = 0
	storing = FALSE
	printed_pre_nulls = FALSE
	first_file = TRUE
	#
	# Parse command line options
	#
	options = ""
	begin_reg = ""
	end_reg = ""
	delimiter = " "
	del_length = 1
	ignore_reg = ""
	omit_reg = ""
	number = 0
	placement = 0
	parse_command()
	}
{
#
# If this is the first file, set the flag, otherwise, reset the environment
#
if (FNR == 1)
	{
	if (first_file) first_file = FALSE
	else reset_new()
	}
#
# If the buffer array has become too large, delete unused entries
#
if (buffer_head > 5000)
	reset_space()
#
# If there is more than one spine present, print an error
#
if ($0 !~ /^!!/ && NF > 1)
	{
	print "context: ERROR: Only one spine is allowed in input (line " FNR ")"\
		 "." > stderr
	exit
	}
#
# If an exclusive interpretation or spine-path terminator occur, process
#
else if ($0 ~ /^\*\*/ || $0 ~ /^\*-$/) process_terminate()
#
# If a comment, tandem interpretation, or null data token occur, process
#
else if ($0 ~ echo) process_echo()
#
# Otherwise, process the given data record according to the options
#
else process_data()
}

############################################################################
#				FUNCTIONS USED IN CONTEXT.AWK
############################################################################

############################################################################
##				Function Parse_command
# This function inspects the array ARGV in order to determine the correct
# options that the user has specified.
#
function parse_command()
	{
	#
	# Delete the DUMMY variable sent from the shell script
	#
	ARGV[1] = ""
	if (ARGV[2] != "null")
		{ begin_reg = ARGV[2] ; options = options "b" }
	if (ARGV[3] != "null")
		{ delimiter = ARGV[3] ; del_length = length(delimiter) }
	if (ARGV[4] != "null")
		{ end_reg = ARGV[4] ; options = options "e" }
	if (ARGV[5] != "null")
		{ ignore_reg = ARGV[5] ; options = options "i" }
	if (ARGV[6] != "null")
		{ number = ARGV[6] ; options = options "n" }
	if (ARGV[7] != "null")
		{ omit_reg = ARGV[7] ; options = options "o" }
	if (ARGV[8] != "null")
		{ placement = ARGV[8] ; options = options "p" }
	if (options ~ /[be]/ && options ~ /[npi]/)
		{
		print "context: ERROR: Cannot specify -b or -e options with "\
			 "-n, -p, or -i options." > stderr
		print USAGE > stderr
		exit
		}
	if (options ~ /n/)
		{
		if (number < 2)
			{
			print "context: ERROR: Number for -n option must be greater "\
				 "than 1." > stderr
			exit
			}
		if (options ~ /p/)
			{
			if (placement < 1)
				{
				print "context: ERROR: Number for -p option must be "\
					 "positive." > stderr
				exit
				}
			else if (placement > number - 1)
				{
				print "context: ERROR: -p number must be at most one less "\
					 "than -n number." > stderr
				exit
				}
			}
		}
	ARGV[2] = ARGV[3] = ARGV[4] = ARGV[5] = ARGV[6] = ARGV[7] = ARGV[8] = ""
	}

##############################################################################
##					Function reset_new
#	This function resets the environment for each new file after the first one.
#
function reset_new(   i)
	{
	buffer_head = 0
	buffer_tail = 1
	data_count = 0
	storing = FALSE
	printed_pre_nulls = FALSE
	for (i in buffer) delete buffer[i]
	}

##############################################################################
##					Function reset_space
#	This function resets the environment when too many lines have been stored
# in the buffer array
#
function reset_space(   i,counter,old_head)
	{
	counter = 0
	old_head = buffer_head
	#
	# Move buffer entries back to the beginning of the array and delete
	# the old entries from the array
	#
	if (buffer_tail <= buffer_head)
		{
		for (i = buffer_tail; i <= buffer_head; i += 1)
			buffer[++counter] = buffer[i]
		buffer_tail = 1
		buffer_head = counter
		for (i = counter + 1; i <= buffer_head; i += 1) delete buffer[i]
		}
	#
	# If buffer_tail is greater than buffer_head, then no entries need to
	# be saved
	#
	else
		{
		buffer_tail = 1
		buffer_head = 0
		for (i = 1; i <= buffer_head; i += 1) delete buffer[i]
		}
	}

##############################################################################
##					Function process_terminate
#	This function processes those records containing a spine-path terminator
# or an exclusive interpretation.
#
function process_terminate(    i,null_tokens)
	{
	#
	# If currently storing records
	#
	if (storing)
		{
		#
		# If -n option was specified
		#
		if (options ~ /n/)
			{
			#
			# If not enough data tokens exist, then print out null tokens
			#
			if (data_count < number)
				{
				for (i = buffer_tail; i <= buffer_head; i += 1)
					{
					if (buffer[i] ~ echo) print buffer[i]
					else print "."
					}
				}
			#
			# Otherwise, print according to the entries in the array and
			# print extra null tokens according to the -p option if specified
			#
			else
				{
				null_tokens = number-1-placement
				for (i = buffer_tail + 1; i <= buffer_head; i += 1)
					{
					if (buffer[i] ~ echo)
						print buffer[i]
					else if (options ~ /o/ && buffer[i] ~ omit_reg)
						print "."
					else if (options ~ /i/ && buffer[i] ~ ignore_reg)
						print "."
					else if (null_tokens > 0)
						{
						print "."
						null_tokens -= 1
						}
					}
				}
			#
			# Reset the variables for the -n option
			#
			data_count = 0
			buffer_tail = buffer_head + 1
			storing = FALSE
			printed_pre_nulls = FALSE
			}
		#
		# Otherwise, for the other options, simply flush the buffer contents
		#
		else if (options ~ /[be]/) flush_buffer()
		}
	#
	# If not storing in buffer array, simply print the record
	#
	print $0
	}

###############################################################################
##					Function flush_buffer
#	This function ouputs the current contents of buffer for the -b and -e
# options.
#
function flush_buffer(   i,output_token,printed)
	{
	printed = FALSE
	output_token = ""
	#
	# Move through the appropriate section of the buffer array and construct
	# the output token of amalgamated data tokens
	#
	for (i = buffer_tail; i <= buffer_head; i += 1)
		{
		if (buffer[i] ~ echo) continue
		else if (options ~ /o/ && buffer[i] ~ omit_reg) continue
		else output_token = output_token buffer[i] delimiter
		}
	output_token = substr(output_token,1,length(output_token) - del_length)
	#
	# Again, loop through the array and print the appropriate tokens at
	# each position
	#
	for (i = buffer_tail; i <= buffer_head; i += 1)
		{
		if (buffer[i] ~ echo) print buffer[i]
		else if (options ~ /o/ && buffer[i] ~ omit_reg) print "."
		else if (printed) print "."
		else { print output_token ; printed = TRUE }
		}
	buffer_tail = buffer_head + 1
	}

##############################################################################
##					Function process_echo
#	This function processes those records containing tokens that should be
# output where they were found in the input.
#
function process_echo()
	{
	if (storing) buffer[++buffer_head] = $0
	else print $0
	}

#############################################################################
##					Function process_data
#	This function processes the data tokens according to the specified options
#
function process_data()
	{
	if (options ~ /n/) process_noption()
	else if (options ~ /b/) process_boption()
	else process_eoption()
	}

############################################################################
##					Function process_noption
#	This function processes data records for the -n option
#
function process_noption(   i)
	{
	#
	# If not currently storing, begin storing
	#
	if (!storing) storing = TRUE
	#
	# If this record should be ignored or omitted, store it in the buffer
	#
	if ((options ~ /i/ && $0 ~ ignore_reg) || (options ~ /o/ && $0 ~ omit_reg))
		{
		if (options ~ /o/) buffer[++buffer_head] = "."	# NEW
		else buffer[++buffer_head] = $0
		}
	else
		{
		#
		# If the data count is not yet full, add this record to the buffer
		# and increment the count
		#
		if (data_count < number)
			{
			buffer[++buffer_head] = $0
			data_count += 1
			#
			# If the data_count is now full, process the data tokens
			#
			if (data_count == number)
				{
				#
				# If the pre-nulls (according to -p option) have not been
				# printed, print them now
				#
				if (!printed_pre_nulls)
					{
					for (i = 1; i <= placement; i += 1) print "."
					printed_pre_nulls = TRUE
					}
				#
				# Print the contents of the buffer and continue to print
				# until the buffer tail is pointing to a data record that is
				# not to be ignored or omitted
				#
				print_buffer()
				if (options ~ /i/ || options ~ /o/)
					{
					while ((options ~ /o/ \
							&& buffer[buffer_tail] ~ omit_reg) \
							|| (options ~ /i/ \
							&& buffer[buffer_tail] ~ ignore_reg))
						{
						buffer_tail += 1
						print_buffer()
						}
					}
				}
			}
		#
		# If the data count is already full, then we just finished printing a
		# set of records and we can do so again by storing the new token and
		# incrementing buffer_tail as well.  Print the buffer again as above
		#
		else if (data_count == number)
			{
			buffer_tail += 1
			buffer[++buffer_head] = $0
			print_buffer()
			if (options ~ /i/ || options ~ /o/)
				{
				while ((options ~ /o/ && buffer[buffer_tail] ~ omit_reg) \
				     || (options ~ /i/ && buffer[buffer_tail] ~ ignore_reg))
					{
					buffer_tail += 1
					print_buffer()
					}
				}
			}
		}
	}

############################################################################
##					Function print_buffer
#	This function prints the buffer for the -n option
#
function print_buffer(   i,output_token)
	{
	output_token = ""
	#
	# Print out any 'non-data' tokens
	#
	i = buffer_tail
	while (buffer[i] ~ echo) { print buffer[i] ; i += 1 }
	buffer_tail = i
	#
	# Loop through the buffer and construct the output token
	#
	for (i = buffer_tail; i <= buffer_head; i += 1)
		{
		if (!(options ~ /o/ && buffer[i] ~ omit_reg) && (buffer[i] !~ echo))
			output_token = output_token buffer[i] delimiter
		}
	#
	# Print the output token
	#
	output_token = substr(output_token,1,length(output_token) - del_length)
	print output_token
	}

############################################################################
##					Function process_boption
#	This function processes data records for the -b option
#
function process_boption()
	{
	#
	# -e option only takes precedence if this record contains a token that is
	# matched by the end regular expression
	#
	if (options ~ /e/ && $0 ~ end_reg) process_eoption()
	#
	# If storing and a 'begin' is found, print buffer and start again
	#
	else if (storing && $0 ~ begin_reg)
		{
		flush_buffer()
		buffer[++buffer_head] = $0
		}
	#
	# If simply storing, then store this record
	#
	else if (storing) buffer[++buffer_head] = $0
	#
	# If this is the first 'begin', then store it and set the flag
	#
	else if ($0 ~ begin_reg)
		{
		buffer[++buffer_head] = $0
		storing = TRUE
		}
	#
	# Otherwise, if this token should be omitted, print null, otherwise print
	#
	else
		{
		if (options ~ /o/ && $0 ~ omit_reg) print "."
		else print $0
		}
	}

############################################################################
##					Function process_eoption
#	This function processes data records for the -e option
#
function process_eoption()
	{
	#
	# If currently storing and this is an end token, store it in the buffer,
	# print the buffer, and set the flag depending on -b option
	#
	if (storing && $0 ~ end_reg)
		{
		buffer[++buffer_head] = $0
		flush_buffer()
		if (options ~ /b/) storing = FALSE
		}
	#
	# If simply storing, then store this record
	#
	else if (storing) buffer[++buffer_head] = $0
	#
	# If this record has an 'end' token, then print it and begin storing if
	# the -b option was not specified
	#
	else if ($0 ~ end_reg)
		{
		print $0
		if (options !~ /b/) storing = TRUE
		}
	#
	# Otherwise, simply store the record and set the storing flag
	#
	else
		{
		storing = TRUE
		buffer[++buffer_head] = $0
		}
	}
